core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java

```diff
@@ -181,15 +182,21 @@ public RunT loadBuild(File dir) throws IOException {
     public final synchronized RunT newBuild() throws IOException {
         try {
             RunT lastBuild = getBuildClass().getConstructor(asJob().getClass()).newInstance(asJob());
+            var rootDir = lastBuild.getRootDir().toPath();
+            if (Files.isDirectory(rootDir)) {
+               LOGGER.warning(() -> "JENKINS-23152: " + rootDir + " already existed; will not overwrite with " + lastBuild + " but will create a fresh build #" + asJob().getNextBuildNumber());
+               return newBuild();
```

Comment:
Yes this is (almost) tail-recursion and could be rewritten to use a loop. OTOH if there is actually a list of thousands of builds on disk newer than the purported `nextBuildNumber`, a `StackOverflowError` is arguably no worse than the previous behavior, and arguably preferable to going into an endless loop. Could tweak to be a bounded loop etc.